From 48af7e5547c6d5d4f1d76e5339a94de5a6a44028 Mon Sep 17 00:00:00 2001
From: Li Nan <linan122@huawei.com>
Date: Thu, 18 May 2023 19:52:30 +0800
Subject: [PATCH 024/263] [Huawei] md/raid10: prioritize read from rdev instrad
 of replacement

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188642

--------------------------------

We have recently been troubled by an issue: use fsstress to test file
system on RAID, constantly inject faulty to disk and plug/unplug the
disks under RAID. We always keeping two disks still to ensure data
integrity. During the test, the file system became read-only.

We have solved many related IO loss issues, but still cannot completely
solve this problem. Since every bug in the read-write process can trigger
the problem, and we found that the problem only occurs when there is a
replacement, we try to avoid reading data from the replacement to prevent
the problem from happening. This is just a temporary solution.

Signed-off-by: Li Nan <linan122@huawei.com>
---
 drivers/md/raid10.c | 32 ++++++++++++++++++++++++--------
 1 file changed, 24 insertions(+), 8 deletions(-)

diff --git a/drivers/md/raid10.c b/drivers/md/raid10.c
index b226f8b4f6e6..66a5797619f1 100644
--- a/drivers/md/raid10.c
+++ b/drivers/md/raid10.c
@@ -708,22 +708,23 @@ static sector_t raid10_find_virt(struct r10conf *conf, sector_t sector, int dev)
  */
 static struct md_rdev *read_balance(struct r10conf *conf,
 				    struct r10bio *r10_bio,
-				    int *max_sectors)
+				    int *max_sectors, bool rdev_prio)
 {
 	const sector_t this_sector = r10_bio->sector;
 	int disk, slot;
-	int sectors = r10_bio->sectors;
+	int sectors;
 	int best_good_sectors;
 	sector_t new_distance, best_dist;
 	struct md_rdev *best_dist_rdev, *best_pending_rdev, *rdev = NULL;
 	int do_balance;
 	int best_dist_slot, best_pending_slot;
-	bool has_nonrot_disk = false;
+	bool has_nonrot_disk;
 	unsigned int min_pending;
 	struct geom *geo = &conf->geo;
 
 	raid10_find_phys(conf, r10_bio);
 	rcu_read_lock();
+retry:
 	best_dist_slot = -1;
 	min_pending = UINT_MAX;
 	best_dist_rdev = NULL;
@@ -731,6 +732,8 @@ static struct md_rdev *read_balance(struct r10conf *conf,
 	best_dist = MaxSector;
 	best_good_sectors = 0;
 	do_balance = 1;
+	sectors = r10_bio->sectors;
+	has_nonrot_disk = false;
 	clear_bit(R10BIO_FailFast, &r10_bio->state);
 	/*
 	 * Check if we can balance. We can balance on the whole
@@ -755,7 +758,10 @@ static struct md_rdev *read_balance(struct r10conf *conf,
 		if (r10_bio->devs[slot].bio == IO_BLOCKED)
 			continue;
 		disk = r10_bio->devs[slot].devnum;
-		rdev = rcu_dereference(conf->mirrors[disk].replacement);
+		if (!rdev_prio)
+			rdev = rcu_dereference(conf->mirrors[disk].replacement);
+		else
+			rdev = NULL;
 		if (rdev == NULL || test_bit(Faulty, &rdev->flags) ||
 		    test_bit(WantRemove, &rdev->flags) ||
 		    r10_bio->devs[slot].addr + sectors >
@@ -779,8 +785,13 @@ static struct md_rdev *read_balance(struct r10conf *conf,
 		dev_sector = r10_bio->devs[slot].addr;
 		if (is_badblock(rdev, dev_sector, sectors,
 				&first_bad, &bad_sectors)) {
-			if (best_dist < MaxSector)
-				/* Already have a better slot */
+			if (best_dist < MaxSector || rdev_prio)
+				/*
+				 * Already have a better slot.
+				 * Or prevent io loss while the data are only
+				 * readable from replacement (rdev isn't NULL
+				 * and will not retry).
+				 */
 				continue;
 			if (first_bad <= dev_sector) {
 				/* Cannot read here.  If this is the
@@ -853,6 +864,11 @@ static struct md_rdev *read_balance(struct r10conf *conf,
 		}
 	}
 
+	if (slot < 0 && rdev_prio) {
+		rdev_prio = false;
+		goto retry;
+	}
+
 	if (slot >= 0) {
 		atomic_inc(&rdev->nr_pending);
 		r10_bio->read_slot = slot;
@@ -1184,7 +1200,7 @@ static void raid10_read_request(struct mddev *mddev, struct bio *bio,
 	}
 
 	regular_request_wait(mddev, conf, bio, r10_bio->sectors);
-	rdev = read_balance(conf, r10_bio, &max_sectors);
+	rdev = read_balance(conf, r10_bio, &max_sectors, true);
 	if (!rdev) {
 		if (err_rdev) {
 			pr_crit_ratelimited("md/raid10:%s: %s: unrecoverable I/O read error for block %llu\n",
@@ -4569,7 +4585,7 @@ static sector_t reshape_request(struct mddev *mddev, sector_t sector_nr,
 	r10_bio->sector = sector_nr;
 	set_bit(R10BIO_IsReshape, &r10_bio->state);
 	r10_bio->sectors = last - sector_nr + 1;
-	rdev = read_balance(conf, r10_bio, &max_sectors);
+	rdev = read_balance(conf, r10_bio, &max_sectors, false);
 	BUG_ON(!test_bit(R10BIO_Previous, &r10_bio->state));
 
 	if (!rdev) {
-- 
2.17.1

